# Matriks Invers
Matriks invers adalah kebalikan dari suatu matriks. jadi jika dilambangkan suatu matriks adalah matriks $A$ maka kebalikan atau invers dari matriks $A$ adalah matriks $A^{-1}$. Matriks $A^{-1}$ dapat dikatakan invers dari matriks $A$ jika memenuhi syarat apabila matriks $A$ dikalikan dengan matriks $A^{-1}$ menghasilkan matriks $I$ yang mana matriks $I$ adalah matriks identitas

## Sifat sifat Matriks Invers 
- Invers dari matriks Invers adalah matriks itu sendiri, contoh invers dari matrik $A^{-1}$ adalah matriks $A$ itu sendiri
- Perkalian matriks dengan inversnya akan menghasilkan matriks identitas, contoh $A \times A^{-1} = A^{-1} \times A = I $ 
- Invers dari perkalian dua matriks adalah hasil dari perkalian inversnya dalam urutan terbalik, contoh $(AB)^{-1} = B^{-1}A^{-1} $
- Invers dari matriks transpose adalah transpose dari matriks invers, contoh $(A^T)^{-1} = (A^{-1})^T$
- Invers dari matriks identitas adalah matriks identitas itu sendiri, contoh $I \times I^{-1} = I$

## Mencari Matriks Invers dengan metode Gauss-Jordan
Metode Gauss Jordan adalah adalah salah satu metode eliminasi yang digunakan untuk menyelesaikan sistem persamaan linear dan mendapatkan solusi yang diberikan dengan cara diubah menjadi augmented matriks atau matriks yang diperluas yang a disederhanakan menjadi bentuk baris yang tereduksi 
contohnya diberikan matriks $A_{3 \times 3}$

$A=
\begin{bmatrix}
1 & 2 & 3 \\
0 & 1 & 4 \\
5 & 6 & 0
\end{bmatrix}$

Kemudian buat augmented matrix atau perluasan dari matriks $A$ dengan menambahkan matriks identiitas sehingga membentuk matriks $AI$ 

$AI =
\begin{bmatrix}
1 & 2 & 3 & 1 & 0 & 0 \\
0 & 1 & 4 & 0 & 1 & 0 \\
5 & 6 & 0 & 0 & 0 & 1 
\end{bmatrix}$

Lakukan operasi sehingga menghasilkan matriks $IA^{-1}$

In [1]:
# import library numpy
import numpy as np

def RowSwap(A,k,l):
# =============================================================================
#     A adalah sebuah NumPy array.  RowSwap akan menghasilkan duplicate dari 
#     array dengan baris k dan l di tukar
# =============================================================================
    m = A.shape[0]  # m adalah nomor baris di A
    n = A.shape[1]  # n adalah nomor kolom di A

    B = np.copy(A).astype('float64')

    for j in range(n):
        temp = B[k][j]
        B[k][j] = B[l][j]
        B[l][j] = temp

    return B

def RowScale(A,k,scale):
# =============================================================================
#     A adalah sebuah NumPy array.  RowScale akan menghasilkan duplicate dari 
#     array dengan baris k di kali dengan skalar bukan 0
# =============================================================================
    m = A.shape[0]  # m adalah nomor baris di A
    n = A.shape[1]  # n adalah nomor kolom di A

    B = np.copy(A).astype('float64')

    for j in range(n):
        B[k][j] *= scale

    return B

def RowAdd(A,k,l,scale):
# =============================================================================
#     A adalah sebuah NumPy array.  RowAdd akan menghasilkan duplicate dari 
#     array dengan baris k akan di kali dengan 'scale' bukan 0. lalu nilai
#     baris l akan di tambah dengan nilai baris k yang sudah dikalikan  
# =============================================================================
    m = A.shape[0]  # m adalah nomor baris di A
    n = A.shape[1]  # n adalah nomor kolom di A

    B = np.copy(A).astype('float64')

    for j in range(n):
        B[l][j] += B[k][j]*scale

    return B

Gunakan fungsi **RowAdd**, **Rowscale**, dan **RowSwap** untuk melakukan operasi

In [2]:
AI = np.array([[1,2,3,1,0,0],[0,1,4,0,1,0],[5,6,0,0,0,1]])
print(AI)

[[1 2 3 1 0 0]
 [0 1 4 0 1 0]
 [5 6 0 0 0 1]]


In [3]:
AI1 = RowAdd(AI,0,2,-5)
print(AI1)

[[  1.   2.   3.   1.   0.   0.]
 [  0.   1.   4.   0.   1.   0.]
 [  0.  -4. -15.  -5.   0.   1.]]


In [4]:
AI2 = RowAdd(AI1,1,2,4)
print(AI2)

[[ 1.  2.  3.  1.  0.  0.]
 [ 0.  1.  4.  0.  1.  0.]
 [ 0.  0.  1. -5.  4.  1.]]


In [5]:
AI3 = RowAdd(AI2,2,1,-4)
print(AI3)

[[  1.   2.   3.   1.   0.   0.]
 [  0.   1.   0.  20. -15.  -4.]
 [  0.   0.   1.  -5.   4.   1.]]


In [6]:
AI4 = RowAdd(AI3,2,0,-3)
print(AI4)

[[  1.   2.   0.  16. -12.  -3.]
 [  0.   1.   0.  20. -15.  -4.]
 [  0.   0.   1.  -5.   4.   1.]]


In [7]:
AI5 = RowAdd(AI4,1,0,-2)
print(AI5)

[[  1.   0.   0. -24.  18.   5.]
 [  0.   1.   0.  20. -15.  -4.]
 [  0.   0.   1.  -5.   4.   1.]]


Dari operasi diatas bisa didapatkan hasil invers dari matriks $A$ adalah

$A^{-1} = 
\begin{bmatrix}
-24 & 18 & 5 \\
20 & -15 & -4 \\
-5 & 4 & 1
\end{bmatrix}
$

Sekarang kita cek apakah matriks $A$ dikalikan dengan matrik $A^{-1}$ sama dengan matriks $I$ atau matriks identitas

In [8]:
A = np.array([[1,2,3],[0,1,4],[5,6,0]])
AI6 = np.array([[-24,18,5],[20,-15,-4],[-5,4,1]])
print(A)
print(AI6)

[[1 2 3]
 [0 1 4]
 [5 6 0]]
[[-24  18   5]
 [ 20 -15  -4]
 [ -5   4   1]]


In [9]:
I = np.dot(A,AI6)
print(I)

[[1 0 0]
 [0 1 0]
 [0 0 1]]


## Penyelesaian sistem persamaan linier dengan metode Gauss-Jordan (Persamaan 4 variabel)

Diberi contoh persamaan linier 4 variabel sebagai berikut
- $ 2x_1 + x_2 + x_3 + 3x_4 = 6$ 
- $ x_1 + 3x_2 + 2x_3 + 5x_4 = 10$ 
- $ 5x_1 + 4x_2 + 3x_3 + x_4 = 11$ 
- $ 6x_1 + 5x_2 + 4x_3 + 2x_4 = 14$ 

Kita buat persamaan $AX = B$ dalam bentuk matriks

$AX = B \\
\begin{bmatrix}
2 & 1 & 1 & 3 \\
1 & 3 & 2 & 5 \\
5 & 4 & 3 & 1 \\
6 & 5 & 4 & 2 
\end{bmatrix}$
$\begin{bmatrix}
x_1 \\ 
x_2 \\ 
x_3 \\
x_4 \\
\end{bmatrix}$
$=$
$\begin{bmatrix}
6 \\ 
10 \\
11 \\
14  
\end{bmatrix}$

Setelah itu, kita ubah menjadi augmented matrix yang ditambahkan dengan matriks identitas. Kita lambangkan sebagai matriks $AI$

$AI =
\begin {bmatrix}
2 & 1 & 1 & 3 & 1 & 0 & 0 & 0 \\
1 & 3 & 2 & 5 & 0 & 1 & 0 & 0 \\
5 & 4 & 3 & 1 & 0 & 0 & 1 & 0 \\
6 & 5 & 4 & 2 & 0 & 0 & 0 & 1 \\
\end {bmatrix}$

Setelah itu kita operasikan hingga menghasilkan matriks $IA^{-1}$

In [10]:
# import library numpy
import numpy as np

def RowSwap(A,k,l):
# =============================================================================
#     A adalah sebuah NumPy array.  RowSwap akan menghasilkan duplicate dari 
#     array dengan baris k dan l di tukar
# =============================================================================
    m = A.shape[0]  # m adalah nomor baris di A
    n = A.shape[1]  # n adalah nomor kolom di A

    B = np.copy(A).astype('float64')

    for j in range(n):
        temp = B[k][j]
        B[k][j] = B[l][j]
        B[l][j] = temp

    return B

def RowScale(A,k,scale):
# =============================================================================
#     A adalah sebuah NumPy array.  RowScale akan menghasilkan duplicate dari 
#     array dengan baris k di kali dengan skalar bukan 0
# =============================================================================
    m = A.shape[0]  # m adalah nomor baris di A
    n = A.shape[1]  # n adalah nomor kolom di A

    B = np.copy(A).astype('float64')

    for j in range(n):
        B[k][j] *= scale

    return B

def RowAdd(A,k,l,scale):
# =============================================================================
#     A adalah sebuah NumPy array.  RowAdd akan menghasilkan duplicate dari 
#     array dengan baris k akan di kali dengan 'scale' bukan 0. lalu nilai
#     baris l akan di tambah dengan nilai baris k yang sudah dikalikan  
# =============================================================================
    m = A.shape[0]  # m adalah nomor baris di A
    n = A.shape[1]  # n adalah nomor kolom di A

    B = np.copy(A).astype('float64')

    for j in range(n):
        B[l][j] += B[k][j]*scale

    return B

In [11]:
BI = np.array([[2,1,1,3,1,0,0,0],[1,3,2,5,0,1,0,0],[5,4,3,1,0,0,1,0],[6,5,4,2,0,0,0,1]])
print(BI)

[[2 1 1 3 1 0 0 0]
 [1 3 2 5 0 1 0 0]
 [5 4 3 1 0 0 1 0]
 [6 5 4 2 0 0 0 1]]


In [12]:
BI1 = RowAdd(BI,0,1,-3)
print(BI1)

[[ 2.  1.  1.  3.  1.  0.  0.  0.]
 [-5.  0. -1. -4. -3.  1.  0.  0.]
 [ 5.  4.  3.  1.  0.  0.  1.  0.]
 [ 6.  5.  4.  2.  0.  0.  0.  1.]]


In [13]:
BI2 = RowAdd(BI1,0,2,-4)
print(BI2)

[[  2.   1.   1.   3.   1.   0.   0.   0.]
 [ -5.   0.  -1.  -4.  -3.   1.   0.   0.]
 [ -3.   0.  -1. -11.  -4.   0.   1.   0.]
 [  6.   5.   4.   2.   0.   0.   0.   1.]]


In [14]:
BI3 = RowAdd(BI2,0,3,-5)
print(BI3)

[[  2.   1.   1.   3.   1.   0.   0.   0.]
 [ -5.   0.  -1.  -4.  -3.   1.   0.   0.]
 [ -3.   0.  -1. -11.  -4.   0.   1.   0.]
 [ -4.   0.  -1. -13.  -5.   0.   0.   1.]]


In [15]:
BI4 = RowScale(BI3,1,-1)
print(BI4)

[[  2.   1.   1.   3.   1.   0.   0.   0.]
 [  5.  -0.   1.   4.   3.  -1.  -0.  -0.]
 [ -3.   0.  -1. -11.  -4.   0.   1.   0.]
 [ -4.   0.  -1. -13.  -5.   0.   0.   1.]]


In [16]:
BI5 = RowScale(BI4,2,-1)
print(BI5)

[[  2.   1.   1.   3.   1.   0.   0.   0.]
 [  5.  -0.   1.   4.   3.  -1.  -0.  -0.]
 [  3.  -0.   1.  11.   4.  -0.  -1.  -0.]
 [ -4.   0.  -1. -13.  -5.   0.   0.   1.]]


In [17]:
BI6 = RowScale(BI5,3,-1)
print(BI6)

[[ 2.  1.  1.  3.  1.  0.  0.  0.]
 [ 5. -0.  1.  4.  3. -1. -0. -0.]
 [ 3. -0.  1. 11.  4. -0. -1. -0.]
 [ 4. -0.  1. 13.  5. -0. -0. -1.]]


In [18]:
BI7= RowAdd(BI6,1,0,-1)
print(BI7)

[[-3.  1.  0. -1. -2.  1.  0.  0.]
 [ 5. -0.  1.  4.  3. -1. -0. -0.]
 [ 3. -0.  1. 11.  4. -0. -1. -0.]
 [ 4. -0.  1. 13.  5. -0. -0. -1.]]


In [19]:
BI8 = RowAdd(BI7,1,2,-1)
print(BI8)

[[-3.  1.  0. -1. -2.  1.  0.  0.]
 [ 5. -0.  1.  4.  3. -1. -0. -0.]
 [-2.  0.  0.  7.  1.  1. -1.  0.]
 [ 4. -0.  1. 13.  5. -0. -0. -1.]]


In [20]:
BI9 = RowAdd(BI8,1,3,-1)
print(BI9)

[[-3.  1.  0. -1. -2.  1.  0.  0.]
 [ 5. -0.  1.  4.  3. -1. -0. -0.]
 [-2.  0.  0.  7.  1.  1. -1.  0.]
 [-1.  0.  0.  9.  2.  1.  0. -1.]]


In [21]:
BI10 = RowAdd(BI9,3,0,-3)
print(BI10)

[[  0.   1.   0. -28.  -8.  -2.   0.   3.]
 [  5.  -0.   1.   4.   3.  -1.  -0.  -0.]
 [ -2.   0.   0.   7.   1.   1.  -1.   0.]
 [ -1.   0.   0.   9.   2.   1.   0.  -1.]]


In [22]:
BI11 = RowAdd(BI10,3,1,5)
print(BI11)

[[  0.   1.   0. -28.  -8.  -2.   0.   3.]
 [  0.   0.   1.  49.  13.   4.   0.  -5.]
 [ -2.   0.   0.   7.   1.   1.  -1.   0.]
 [ -1.   0.   0.   9.   2.   1.   0.  -1.]]


In [23]:
BI12 = RowAdd(B11,3,2,-2)
print(BI12)

NameError: name 'B11' is not defined

In [ ]:
BI13 = RowScale(B12,2,-1/11)
print(BI13)

[[  0.           1.           0.         -28.          -8.
   -2.           0.           3.        ]
 [  0.           0.           1.          49.          13.
    4.           0.          -5.        ]
 [ -0.          -0.          -0.           1.           0.27272727
    0.09090909   0.09090909  -0.18181818]
 [ -1.           0.           0.           9.           2.
    1.           0.          -1.        ]]


In [ ]:
BI14 =  RowAdd(BI13,2,0,28)
print(BI14)

[[ 0.          1.          0.          0.         -0.36363636  0.54545455
   2.54545455 -2.09090909]
 [ 0.          0.          1.         49.         13.          4.
   0.         -5.        ]
 [-0.         -0.         -0.          1.          0.27272727  0.09090909
   0.09090909 -0.18181818]
 [-1.          0.          0.          9.          2.          1.
   0.         -1.        ]]


In [ ]:
BI15 = RowAdd(BI14,2,1,-49)
print(BI15)

[[ 0.          1.          0.          0.         -0.36363636  0.54545455
   2.54545455 -2.09090909]
 [ 0.          0.          1.          0.         -0.36363636 -0.45454545
  -4.45454545  3.90909091]
 [-0.         -0.         -0.          1.          0.27272727  0.09090909
   0.09090909 -0.18181818]
 [-1.          0.          0.          9.          2.          1.
   0.         -1.        ]]


In [ ]:
BI16 = RowAdd(BI15,2,3,-9)
print(BI16)

[[ 0.          1.          0.          0.         -0.36363636  0.54545455
   2.54545455 -2.09090909]
 [ 0.          0.          1.          0.         -0.36363636 -0.45454545
  -4.45454545  3.90909091]
 [-0.         -0.         -0.          1.          0.27272727  0.09090909
   0.09090909 -0.18181818]
 [-1.          0.          0.          0.         -0.45454545  0.18181818
  -0.81818182  0.63636364]]


In [ ]:
BI17 = RowScale(BI16,3,-1)
print(BI17)

[[ 0.          1.          0.          0.         -0.36363636  0.54545455
   2.54545455 -2.09090909]
 [ 0.          0.          1.          0.         -0.36363636 -0.45454545
  -4.45454545  3.90909091]
 [-0.         -0.         -0.          1.          0.27272727  0.09090909
   0.09090909 -0.18181818]
 [ 1.         -0.         -0.         -0.          0.45454545 -0.18181818
   0.81818182 -0.63636364]]


In [ ]:
BI18 = RowSwap(BI17,3,0)
print(BI18)

[[ 1.         -0.         -0.         -0.          0.45454545 -0.18181818
   0.81818182 -0.63636364]
 [ 0.          0.          1.          0.         -0.36363636 -0.45454545
  -4.45454545  3.90909091]
 [-0.         -0.         -0.          1.          0.27272727  0.09090909
   0.09090909 -0.18181818]
 [ 0.          1.          0.          0.         -0.36363636  0.54545455
   2.54545455 -2.09090909]]


In [ ]:
BI19 = RowSwap(BI18,3,1)
print(BI19)

[[ 1.         -0.         -0.         -0.          0.45454545 -0.18181818
   0.81818182 -0.63636364]
 [ 0.          1.          0.          0.         -0.36363636  0.54545455
   2.54545455 -2.09090909]
 [-0.         -0.         -0.          1.          0.27272727  0.09090909
   0.09090909 -0.18181818]
 [ 0.          0.          1.          0.         -0.36363636 -0.45454545
  -4.45454545  3.90909091]]


In [ ]:
BI20 = RowSwap(BI19,2,3)
print(BI20)

[[ 1.         -0.         -0.         -0.          0.45454545 -0.18181818
   0.81818182 -0.63636364]
 [ 0.          1.          0.          0.         -0.36363636  0.54545455
   2.54545455 -2.09090909]
 [ 0.          0.          1.          0.         -0.36363636 -0.45454545
  -4.45454545  3.90909091]
 [-0.         -0.         -0.          1.          0.27272727  0.09090909
   0.09090909 -0.18181818]]


In [ ]:
B = np.array([[2,1,1,3],[1,3,2,5],[5,4,3,1],[6,5,4,2]])
BS = np.linalg.inv(B)
print(BS)

[[ 0.45454545 -0.18181818  0.81818182 -0.63636364]
 [-0.36363636  0.54545455  2.54545455 -2.09090909]
 [-0.36363636 -0.45454545 -4.45454545  3.90909091]
 [ 0.27272727  0.09090909  0.09090909 -0.18181818]]


$AX = B \\
A^{-1}A X = A^{-1}B
IX = A^{-1}B$


Setelah ketemu invers matriksnya kita operasikan untuk menemukan X dengan mengalikan matriks $A^{-1}$ dengan matriks $B$

$IX = A^{-1}B$

$\begin{bmatrix}
1 & 0 & 0 & 0 \\
0 & 1 & 0 & 0 \\
0 & 0 & 1 & 0 \\
0 & 0 & 0 & 1
\end{bmatrix}$
$\begin{bmatrix}
x_1 \\ 
x_2 \\ 
x_3 \\
x_4 \\
\end{bmatrix}$
$=$
$\begin{bmatrix}
0.45454545 & -0.18181818 & 0.81818182 & -0.63636364 \\
-0.36363636 & 0.54545455 & 2.54545455 & -2.09090909 \\
-0.36363636 & -0.45454545 & -4.45454545 & 3.90909091 \\
0.27272727 & 0.09090909 & 0.09090909 & -0.18181818
\end{bmatrix}$
$\begin{bmatrix}
6 \\ 
10 \\
11 \\
14  
\end{bmatrix}$


Setelah itu kita operasikan menggunakan fungsi **NumPy** pada Python 

In [ ]:
B = np.array ([[6,10,11,14]])
BT = B.transpose() 
X = np.dot(BS,BT)
print(X)

[[ 1.]
 [ 2.]
 [-1.]
 [ 1.]]


Dari operasi diatas bisa kita dapatkan nilai dari $x_1, x_2, x_3, x_4$ adalah

$x_1 = 1 \\
x_2 = 2 \\
x_3 = -1 \\
x_4 = 1$